src/proto/streams/recv.rs

```diff
@@ -411,6 +415,19 @@ impl Recv {
         Ok(())
     }
 
+    pub fn consume_connection_window(&mut self, sz: WindowSize) -> Result<(), RecvError> {
+        if self.flow.window_size() < sz {
+            return Err(RecvError::Connection(Reason::FLOW_CONTROL_ERROR));
+        }
+
+        // Update connection level flow control
+        self.flow.send_data(sz);
+
+        // Track the data as in-flight
+        self.in_flight_data += sz;
```

Comment:
There is an open question of whether we should immediately release this connection capacity. If we don't, since there is no related stream to release the capacity, the connection capacity is essentially "leaked" forever. At the same time, if we locally reset this stream, but the remote just keeps sending us data anyways, giving capacity back would allow them to just keep flooding the connection with this stream we no longer want.

Perhaps if we add in the "for some time" part, by checking a time elapsed, then the eventual connection error would be a fine deterrent and allow us to free the connection window automatically.

Even without this question answered, it is still better that we count this anyway, because: it **is** correct that the window has been used, it **is not** required that we give the capacity back.

Reply:
IMO we should immediately release capacity. the window updates will be sequenced *after* the  RST_STREAM frame, so the remote should not send any more data.

Reply:
Additionally, we won't ever increment the reset stream's window, so even if the remote keeps sending data, the window of that stream will eventually get depleted.